---
title: "Targeted multi-omics single cell seq (mRNA + protein) on MalariaTravellerCohort"
author: "Maximilian Julius Lautenbach"
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    highlight: pygments
abstract: Short project description.
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(dirname(inputFile), paste0("../Paper-",format(Sys.time(), "%d-%m-%Y-%H-%M-%S"),".docx"))) })
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_knit$set(echo = TRUE, 
                     #root.dir = "/Users/maxlau/Projects_MJL/MalariaTraveller_AbSeq",
                     root.dir = getwd(),
                     fig.width = 6, fig.height = 5, warning = FALSE, message = FALSE)
#knitr::opts_chunk$set(cache = TRUE)
result.dir <- paste0("results/",Sys.Date(),"/")
## creates result.dir with date in if not existent
ifelse(isFALSE(dir.exists(paste0("../",result.dir))), dir.create(paste0("../",result.dir),recursive = TRUE),"Result directory for today exists already!")
options(stringsAsFactors = FALSE) 
#renv::use_python(type = "conda", name="scRNAseq2021")
```


# Libraries
```{r message = FALSE}
library(tidyverse)
library(data.table)
library(Seurat)
library(patchwork)
library(randomcoloR)
#source("helper_fun.R")
```

## Load Seurat Object
* Seurat object containing 12 samples, from 4 libraries
* Doublets and undetermined cells, determined by BD SevenBridges pipline, were removed
* Only cells with "min.cells = 3" and "min.features = 10" on RNA level were inclued

```{r message=F}
#data.raw <- readRDS("../data/SeuratObjects/2021-09-21_AbseqCellCall_MalariaTraveller_RhapsodyLibraries_SeuratObeject.rds")

# Load and add BD pipeline determined cell types
#cell_type_exp <- read_csv("../data/SeuratObjects/2021-09-21_AbseqCellCall_MalariaTraveller_RhapsodyLibraries_CellTypeExperimental.csv")
data.raw <- readRDS("../../data/RhapsodyCountFiles/temp/AbSeqCellCalling_MalariaTraveller_RhapsodyLibraries_SeuratObeject.rds")# AbSeqCellCalling_MalariaTraveller_RhapsodyLibraries_SeuratObeject.rds")#RhapsodyCountFiles/temp/2023-12-22AbSeqCellCalling_MalariaTraveller_RhapsodyLibraries_SeuratObeject.rds")#data/SeuratObjects/2021-09-21_AbseqCellCall_MalariaTraveller_RhapsodyLibraries_SeuratObeject.rds")

# Load and add BD pipeline determined cell types
cell_type_exp <- read_csv("../../data/RhapsodyCountFiles/temp/AbSeqCellCalling_MalariaTraveller_RhapsodyLibraries_CellTypeExperimental.csv")#1_AbseqCellCall_MalariaTraveller_RhapsodyLibraries_CellTypeExperimental.csv")

colOrd <- rownames(FetchData(data.raw,"ident"))
RcolData <- cell_type_exp %>% dplyr::slice(match(colOrd, Cell_ID))

#Add to metadata 
data.raw@meta.data <- bind_cols(data.raw@meta.data, RcolData$Cell_Type_Experimental) %>% rename(Cell_Type_Experimental = "...10")
```

## Inspect data structure

```{r}
data.raw
as.data.frame(data.raw@assays$RNA@counts[1:10, 1:2])
as.data.frame(data.raw@assays$ADT@counts[1:10, 1:2])
head(data.raw@meta.data, 10)
```

### Cell numbers per sample

```{r}
data.raw@meta.data %>%
  as.data.table %>% # the resulting md object has one "row" per cell
  rownames_to_column("Cell-ID") %>% 
  group_by(orig.ident,Group,Time) %>% 
  dplyr::count() %>% 
  ggplot(aes(x=orig.ident,y=n, fill=Group)) +
  coord_flip()+
  geom_col() +
  geom_hline(yintercept = 6000) +
  geom_text(aes(label = Time), hjust =1.5) + 
  geom_text(aes(label = n), hjust =5) + 
  scale_fill_manual(values = ENDEMIC_colors)
```

### Cumulative cell numbers per group

```{r}
data.raw@meta.data %>%
  as.data.table %>% # the resulting md object has one "row" per cell
  rownames_to_column("Cell-ID")  %>% 
  group_by(Group,Time) %>% 
  dplyr::count() %>% 
  ggplot(aes(x=Group,y=n, fill=Time)) +
  coord_flip()+
  geom_bar(stat="identity", position="dodge") +
  geom_hline(yintercept = 12000) +
  scale_fill_manual(values=TIME_colors) +
  labs(x="Exposure group",
       y="Total number of cells",
       title = "Cumulative number of cells per group",
       caption = "planned #: 12.000 cells") +
  scale_y_continuous(breaks = )
```

### Numbers of cells per sample
```{r}
table(data.raw@meta.data$Cell_Type_Experimental)
```

### Library/Sample overview
```{r }
feats <- c("nCount_RNA", "nFeature_RNA", "nCount_ADT", "nFeature_ADT")

VlnPlot(data.raw, group.by = "Library", features = feats, pt.size = 0, ncol = 4) + 
  NoLegend()

VlnPlot(data.raw, group.by = "orig.ident", features = feats, pt.size = 0, ncol = 4) + 
  NoLegend()

VlnPlot(data.raw, group.by = "Group", features = feats, pt.size = 0, ncol = 4) + 
  NoLegend()
```

## create new object
```{r}
alldata <- data.raw
```



In the same manner we will calculate the proportion gene expression that comes from ribosomal proteins.

```{r, results='hold'}
# Way1: Doing it using Seurat function
alldata <- PercentageFeatureSet(alldata, "^HLA", col.name = "percent_HLA")

# Way2: Doing it manually
#hla_genes <- rownames(alldata)[grep("^HLA",rownames(alldata))]
#alldata$percent_hla_man <- (colSums(alldata@assays$RNA@counts[hla_genes,]  ) / total_counts_per_cell)*100
```

# Plot QC

Now we can plot some of the QC-features as violin plots.

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=8,fig.width=16}
feats <- c("nFeature_RNA","nCount_RNA","percent_HLA")
VlnPlot(alldata, group.by= "orig.ident", features = feats, pt.size = 0.1,ncol = 2) + NoLegend()
```

### plot the different QC-measures as scatter plots.

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=16}
FeatureScatter(alldata, "nCount_RNA"  , "nFeature_RNA", group.by = "orig.ident", pt.size = .5)
```

***
# Filtering

### Detection-based filtering

A standard approach is to filter cells with low amount of reads as well as genes that are present in at least a certain amount of cells. Here we will only consider cells with at least 10 detected genes and genes need to be expressed in at least 3 cells. Please note that those values are highly dependent on the library preparation method used.

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=4}
selected_c <- WhichCells(alldata, expression = nFeature_RNA > 10)
selected_f <- rownames(alldata)[ Matrix::rowSums(alldata) > 3]

data.filt <- subset(alldata, features=selected_f, cells=selected_c)

## unfiltered data
dim(alldata)
## filtered for at least 10 detected genes and genes at least expressed in 3 cells
dim(data.filt)
```
-> 372 genes in 73,438 cells

### Removal of high gene number and count cells 

Extremely high number of detected genes could indicate doublets. However, depending on the cell type composition in your sample, you may have cells with higher number of genes (and also higher counts) from one cell type. 

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=16}
# skip for now and run DoubletFinder first!

high.det.v3 <- WhichCells(data.filt, expression = nFeature_RNA > 200)
high.det.v2 <- WhichCells(data.filt, expression = nCount_RNA > 5000)

# remove these cells
data.filt <- subset(data.filt, cells=setdiff(WhichCells(data.filt),c(high.det.v2,high.det.v3)))

# check number of cells
dim(data.filt)
```
-> 372 genes in 73,038 cells

Now we can plot the QC-features as violin plots again

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=8,fig.width=16}
VlnPlot(data.filt, group.by= "orig.ident", features = feats, pt.size = 0.1,ncol = 2) + NoLegend()
```

```{r}
#- Calculate highly variable genes, not used, only for information
data.filt <- FindVariableFeatures(object = data.filt, do.plot = TRUE, assay = "RNA", y.cutoff = 0.2)

VariableFeatures(data.filt) %>% length() # 372 variable genes
```



## Normalization

```{r cache = TRUE}
#- RNA Normalization
alldata <- NormalizeData(object = data.filt, assay = 'RNA', normalization.method = 'LogNormalize', scale.factor = 10000)

#- Ab Normalization
alldata <- NormalizeData(object = data.filt, assay = 'ADT', normalization.method = 'CLR')
```
Now we can plot the QC-features as violin plots again

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=8,fig.width=16}
VlnPlot(data.filt, group.by= "orig.ident", features = feats, pt.size = 0.1,ncol = 2) + NoLegend()
```

```{r}
# Compute the relative expression of each gene per cell Use sparse matrix
# operations, if your dataset is large, doing matrix devisions the regular way
# will take a very long time.
library(Matrix)
par(mar = c(4, 8, 2, 1))
C <- data.filt@assays$RNA@counts
C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
boxplot(as.matrix(t(C[most_expressed, ])), cex = 0.1, las = 1, xlab = "% total count per cell", 
    col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
```


# Weighted Nearest Neighbor integration

## Weighted Nearest Neighbor Analysis on multiome
https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html

The simultaneous measurement of multiple modalities, known as multimodal analysis, represents an exciting frontier for single-cell genomics and necessitates new computational methods that can define cellular states based on multiple data types. The varying information content of each modality, even across cells in the same dataset, represents a pressing challenge for the analysis and integration of multimodal datasets. In (Hao*, Hao* et al, Cell 2021), we introduce ‘weighted-nearest neighbor’ (WNN) analysis, an unsupervised framework to learn the relative utility of each data type in each cell, enabling an integrative analysis of multiple modalities.

This vignette introduces the WNN workflow for the analysis of multimodal single-cell datasets. The workflow consists of three steps

Independent preprocessing and dimensional reduction of each modality individually
Learning cell-specific modality ‘weights’, and constructing a WNN graph that integrates the modalities
Downstream analysis (i.e. visualization, clustering, etc.) of the WNN graph
We demonstrate the use of WNN analysis to two single-cell multimodal technologies: CITE-seq and 10x multiome. We define the cellular states based on both modalities, instead of either individual modality.

We first perform pre-processing and dimensional reduction on both assays independently. We use standard normalization, but you can also use SCTransform or any alternative method.



# CCA Integration
## 1. RNA
#### Anchor points of all 12 samples

```{r}
alldata <- data.raw
```

```{r}
DefaultAssay(alldata) <- "RNA"
alldata.list <- SplitObject(alldata, split.by = "orig.ident")

for (i in 1:length(alldata.list)) {
    alldata.list[[i]] <- NormalizeData(alldata.list[[i]], normalization.method = 'LogNormalize',scale.factor = 10000, verbose = FALSE,)
    alldata.list[[i]] <- FindVariableFeatures(alldata.list[[i]], selection.method = "vst", 
        nfeatures = 399, verbose = FALSE)
}

hvgs_per_dataset <- lapply(alldata.list, function(x) {
    x@assays$RNA@var.features
})
venn::venn(hvgs_per_dataset, opacity = 0.4, zcolor = (scales::hue_pal())(3), cexsn = 1, 
    cexil = 1, lwd = 1, col = "white", frame = F, borders = NA)

rna.anchors <- FindIntegrationAnchors(object.list = alldata.list, 
                                          dims = 1:30, 
                                          reduction = "cca",verbose = F)

#We then pass these anchors to the IntegrateData function, which returns a Seurat object.

rna.obj <- IntegrateData(anchorset = rna.anchors, dims = 1:30, verbose=F, new.assay.name = "rna.CCA")
saveRDS(rna.obj, paste0("../../data/RhapsodyCountFiles/temp/","Object_rnaCAA.rds"))

```



## 2. Ab
#### Anchor points of all 12 samples
```{r }
DefaultAssay(alldata) <- "ADT"
alldata.list <- SplitObject(alldata, split.by = "orig.ident")

for (i in 1:length(alldata.list)) {
    alldata.list[[i]] <- NormalizeData(alldata.list[[i]],assay = "ADT", normalization.method = 'CLR',verbose = FALSE)
    #alldata.list[[i]] <- FindVariableFeatures(alldata.list[[i]], selection.method = "vst", nfeatures = 29, verbose = FALSE)
    VariableFeatures( alldata.list[[i]]) <- rownames(alldata[["ADT"]])
}

hvgs_per_dataset <- lapply(alldata.list, function(x) {
    x@assays$ADT@var.features
})
venn::venn(hvgs_per_dataset, opacity = 0.4, zcolor = (scales::hue_pal())(3), cexsn = 1, 
    cexil = 1, lwd = 1, col = "white", frame = F, borders = NA)

adt.anchors <- FindIntegrationAnchors(object.list = alldata.list, 
                                          dims = 1:18, 
                                          verbose = F,
                                          reduction = "cca")


#We then pass these anchors to the IntegrateData function, which returns a Seurat object.

adt.obj <- IntegrateData(anchorset = adt.anchors, verbose = F,dims = 1:18, new.assay.name = "adt.CCA")
saveRDS(adt.obj, paste0("../../data/RhapsodyCountFiles/temp/","Object_adtCAA.rds"))

```

```{r}
int.rna <- readRDS("../../data/RhapsodyCountFiles/temp/Object_rnaCAA.rds")
int.adt <- readRDS("../../data/RhapsodyCountFiles/temp/Object_adtCAA.rds")

DefaultAssay(int.rna)
DefaultAssay(int.adt)
```

## pipeline structure reference
https://github.com/satijalab/seurat/issues/4815


```{r}
set.seed(42)
int.rna <- ScaleData(int.rna, verbose = FALSE)
int.rna <- RunPCA(int.rna, npcs = 30, verbose = FALSE, reduction.name = "rna.cca.pca")
int.rna <- RunUMAP(int.rna, reduction = 'rna.cca.pca',dim=1:30, reduction.name = 'rna.cca.umap', verbose = FALSE)


int.adt <- ScaleData(int.adt, verbose = FALSE)
int.adt <- RunPCA(int.adt, npcs = 18, verbose = FALSE, reduction.name = "adt.cca.pca")
int.adt <- RunUMAP(int.adt, reduction = 'adt.cca.pca',dim=1:18, verbose = FALSE, reduction.name = "adt.cca.umap")

```


## combination
https://github.com/satijalab/seurat/issues/3890

```{r}
int.both <- int.adt

int.both[["rna.CCA"]] <- int.rna[["rna.CCA"]]
int.both[["rna.cca.pca"]] <- int.rna[["rna.cca.pca"]]
int.both[["rna.cca.umap"]] <- int.rna[["rna.cca.umap"]]

int.both@assays
int.both@reductions
```


# Weighted Nearest Neighbor integration
weighted.nn is the weighted KNN. wsnn is the shared nearest neighbors from weighted.nn.
```{r}
# Identify multimodal neighbors. These will be stored in the neighbors slot, 
# and can be accessed using bm[['weighted.nn']]
# The WNN graph can be accessed at bm[["wknn"]], 
# and the SNN graph used for clustering at bm[["wsnn"]]
# Cell-specific modality weights can be accessed at bm$RNA.weight
int.both <- FindMultiModalNeighbors(int.both,
                                   reduction.list = list("rna.cca.pca", "adt.cca.pca"), 
                                   dims.list = list(1:30, 1:18), 
                                   modality.weight.name = "RNA.weight"
                                   )
int.both[['weighted.nn']]
```

```{r}
int.both <- RunUMAP(int.both, nn.name = 'weighted.nn', reduction.name = 'wnn.umap', reduction.key = 'wnnUMAP_')

#Visualise clusters on UMAP:
DimPlot(int.both, reduction= 'wnn.umap')
```

```{r }
DimPlot(int.both, reduction= 'wnn.umap', group.by = c("orig.ident","Group","Time","Cell_Type_Experimental"), label=F) & NoLegend() & plot_annotation(title = 'integrated WNN')
```


```{r}
DimPlot(int.both, reduction= 'adt.cca.umap', group.by = c("orig.ident","Group","Time","Cell_Type_Experimental"), label=F) & NoLegend() & plot_annotation(title = 'integrated (CCA) ADT UMAP')
```


```{r fig.width=4, fig.height=4}
DimPlot(int.both, reduction= 'rna.cca.umap', group.by = c("orig.ident","Group","Time","Cell_Type_Experimental"), label=F) & NoLegend() & plot_annotation(title = 'integrated (CCA) RNA UMAP')
```


## WNN on mRNA integrated data only

```{r}
int.rna
DefaultAssay(int.rna) <- "ADT"

int.rna <- ScaleData(int.rna, verbose = FALSE) %>% NormalizeData(assay = "ADT", normalization.method = 'CLR',verbose = FALSE) %>%  FindVariableFeatures(selection.method = "vst", nfeatures = 29, verbose = FALSE)
    #VariableFeatures( alldata.list[[i]]) <- rownames(alldata[["ADT"]])
  
VariableFeatures(int.rna)
  
int.rna <- RunPCA(int.rna, npcs = 18, verbose = FALSE, reduction.name = "adt.pca")

int.rna.only <- FindMultiModalNeighbors(int.rna,
                                        reduction.list = list("rna.cca.pca", "adt.pca"), 
                                        dims.list = list(1:30, 1:18), 
                                        modality.weight.name = "RNA.weight")
int.rna.only <- RunUMAP(int.rna.only, nn.name = 'weighted.nn', reduction.name = 'wnn.umap', reduction.key = 'wnnUMAP_')
```

## Clustering on WNN integrated graph
```{r }
# Clustering 
for (res in c(0.05,0.1,0.2,0.25,0.3)) {
    int.rna.only <- FindClusters(int.rna.only, graph.name = "wsnn", algorithm = 3, resolution = res, verbose = FALSE)
}
```


```{r}
p1 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'Group', label = F, repel = TRUE, label.size = 2.5) + NoLegend() + ggplot2::ggtitle(label = "wnn on int (ADT + RNA)")
p2 <- DimPlot(int.rna.only, reduction = 'wnn.umap', group.by = 'Group', label = F, repel = TRUE, label.size = 2.5) + NoLegend()  + ggplot2::ggtitle(label = "wnn on int (RNA)")

p1 + p2 
```


```{r fig.width=6}
p1 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'orig.ident', label = F, repel = TRUE, label.size = 2.5) + NoLegend() + ggplot2::ggtitle(label = "wnn on int (ADT + RNA)")
p2 <- DimPlot(int.rna.only, reduction = 'wnn.umap', group.by = 'orig.ident', label = F, repel = TRUE, label.size = 2.5)  + ggplot2::ggtitle(label = "wnn on int (RNA)")

p1 + p2 
```


## Side by side comparison
```{r fig.width=6, fig.height=2}
p1 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'Group', label = F, repel = TRUE, label.size = 2.5) + NoLegend() + ggplot2::ggtitle(label = "ADT + RNA")
p2 <- DimPlot(int.both, reduction = 'adt.cca.umap', group.by = 'Group', label = F, repel = TRUE, label.size = 2.5) + NoLegend()  + ggplot2::ggtitle(label = "ADT")
p3 <- DimPlot(int.both, reduction = 'rna.cca.umap', group.by = 'Group', label = F, repel = TRUE, label.size = 2.5) + NoLegend()  + ggplot2::ggtitle(label = "RNA")

p1 + p2 + p3
```
```{r fig.width=6, fig.height=2}
p1 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'Time', label = F, repel = TRUE, label.size = 2.5) + NoLegend() + NoAxes() + ggplot2::ggtitle(label = "ADT + RNA")
p2 <- DimPlot(int.both, reduction = 'adt.cca.umap', group.by = 'Time', label = F, repel = TRUE, label.size = 2.5) + NoLegend() + NoAxes() + ggplot2::ggtitle(label = "ADT")
p3 <- DimPlot(int.both, reduction = 'rna.cca.umap', group.by = 'Time', label = F, repel = TRUE, label.size = 2.5) + NoLegend() + NoAxes() + ggplot2::ggtitle(label = "RNA")

p1 + p2 + p3
```



## Clustering on WNN integrated graph
```{r }
# Clustering 
for (res in c(0.05,0.1,0.2,0.25,0.3)) {
    int.both <- FindClusters(int.both, graph.name = "wsnn", algorithm = 3, resolution = res, verbose = FALSE)
}
```


```{r }
p1 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = "wsnn_res.0.1", label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend()
p2 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'Cell_Type_Experimental', label = TRUE, repel = TRUE, label.size = 2.5) 
p1 + p2
```

```{r }
p5 <- FeaturePlot(int.both, features = c("adt_CD16"),
                  reduction = 'wnn.umap', max.cutoff = 3, 
                  cols = c("lightgrey","darkgreen"), ncol = 1)
p6 <- FeaturePlot(int.both, features = c("rna_FCGR3A"), 
                  reduction = 'wnn.umap', max.cutoff = 3, ncol = 1)
p5 | p6
```

```{r fig.width=8, fig.height=3}
p1 <- VlnPlot(int.both, features = "rna.CCA.weight", group.by = 'wsnn_res.0.1', sort = TRUE, pt.size = 0.01) +
  NoLegend()
p2 <- VlnPlot(int.both, features = "adt.CCA.weight", group.by = 'wsnn_res.0.1', sort = TRUE, pt.size = 0.01) +
  NoLegend()

p1 | p2
```

```{r fig.with=8, fig.height=3}
p1 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'wsnn_res.0.1', label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend()
p2 <- DimPlot(int.both, reduction = 'wnn.umap', group.by = 'Cell_Type_Experimental', label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend()
p1 + p2
```

```{r}
##save seurat object
saveRDS(int.both, paste0("../../data/RhapsodyCountFiles/temp/","CCAintRNA_CCAintADT_on_AbseqCellCalling_WNN.rds"))
```

```{r}
saveRDS(int.rna.only, paste0("../../data/RhapsodyCountFiles/temp/","CCAintRNA_on_AbseqCellCalling_data_WNN.rds"))

```
